UPGRADE_GUIDE.md

```diff
@@ -2,7 +2,203 @@
+    - `EasyPost.Exceptions.RetryError` - thrown when an error occurs during a retry
+    - `EasyPost.Exceptions.ServiceUnavailableError` - thrown when the API is unavailable (503 status code)
+    - `EasyPost.Exceptions.SslError` - thrown when the library is unable to connect to the API via SSL
+    - `EasyPost.Exceptions.TimeoutError` - thrown when the API request times out (408 status code)
+    - `EasyPost.Exceptions.UnauthorizedError` - thrown when the API receives an unauthorized request (401 or 403 status
+      code)
+    - `EasyPost.Exceptions.UnexpectedHttpError` - thrown when an unknown HTTP error occurs (unexpected 5xx status code)
+    - `EasyPost.Exceptions.UnknownApiError` - thrown when an unknown API error occurs (unexpected 4xx status code)
+    - `EasyPost.Exceptions.VcrError` - thrown when an error occurs with the VCR (used for testing only)
+- Generic exceptions
+    - `EasyPost.Exceptions.FilteringError` - thrown when an error occurs during filtering (e.g. calculating lowest rate)
+    - `EasyPost.Exceptions.InvalidObjectError` - thrown when an invalid object is being used
+    - `EasyPost.Exceptions.JsonError`
+        - `EasyPost.Exceptions.JsonDeserializationError` - thrown when an error occurs during JSON deserialization
+        - `EasyPost.Exceptions.JsonSerializationError` - thrown when an error occurs during JSON serialization
+        - `EasyPost.Exceptions.JsonNoDataError` - thrown when no data is provided for JSON deserialization
+    - `EasyPost.Exceptions.MissingPropertyError` - thrown when a required property is missing (e.g. `Id` for most
+      objects)
+    - `EasyPost.Exceptions.SignatureVerificationError` - thrown when the signature for a webhook is invalid
+    - `EasyPost.Exceptions.ValidationError`
+        - `EasyPost.Exceptions.MissingParameterError` - thrown when a required parameter is missing
+        - `EasyPost.Exceptions.InvalidParameterError` - thrown when an invalid parameter is passed to a function
+
+Any exception thrown by the EasyPost library will be one of the above types.
+
+Any `EasyPost.Exceptions.ApiError` exception will have the following properties:
+
+- `int? StatusCode` - the HTTP status code returned by the API call (e.g. 404)
+- `string? Code` - the error code returned by the API (e.g. "PARAMETER.INVALID")
+- `List<EasyPost.Error>? Errors` - a list of errors returned by the API (e.g. "Invalid parameter: to_address")
+
+Any `EasyPost.Exceptions.ApiError` exception can be pretty-printed using the `PrettyPrint` string property.
+
+Users can better anticipate exception information through utilities in the `EasyPost.Exceptions.Constants` namespace.
+
+- Check what exception type will be thrown for a given HTTP status code by
+  using `EasyPost.Exceptions.Constants.GetEasyPostExceptionType(HttpStatusCode statusCode)`
+- Check error message strings/templates in `EasyPost.Exceptions.Constants.ErrorMessages`
```

Comment:
Would like some examples of usage for this please

Reply:
What did you have in mind?

Reply:
Mmm re-reading this it's fine.